src/crypto/pubkeyDecryptor.ts

```diff
@@ -0,0 +1,36 @@
+import crypto from "crypto";
+import nacl from "tweetnacl";
+import ed2curve from "ed2curve";
+import {X25519EncryptedData} from "./x25519EncryptedData";
+import {UserPublicKey, UserSecretKey} from "../userKeys";
+
+export class PubkeyDecryptor {
+    static decrypt(data: X25519EncryptedData, decryptorSecretKey: UserSecretKey): Buffer {
+        const ciphertext = Buffer.from(data.ciphertext, 'hex');
+        const edhPubKey = Buffer.from(data.identities.ephemeralPubKey, 'hex');
+        const originatorPubKeyBuffer = Buffer.from(data.identities.originatorPubKey, 'hex');
+        const originatorPubKey = new UserPublicKey(originatorPubKeyBuffer);
+
+        const authMessage = crypto.createHash('sha256').update(
+            Buffer.concat([ciphertext, edhPubKey])
+        ).digest();
+
+        if (!originatorPubKey.verify(authMessage, Buffer.from(data.mac, 'hex'))) {
+            throw new Error("Invalid authentication for encrypted message originator");
```

Comment:
So here we verify the "wrapping" signature.

Reply:
Yes, I don't know if we are over-doing it, but I was thinking of a use case where the "admin" of a contract will publish some encrypted data and would like to have a proof of that. The dev implementing that app could have done it himself but I guess it's nice to have it out-of-the-box.